sync15-adapter/src/util.rs

```diff
@@ -0,0 +1,135 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::convert::From;
+use std::time::Duration;
+use std::{fmt, num};
+use std::str::FromStr;
+use openssl;
+use base64;
+
+pub fn base16_encode(bytes: &[u8]) -> String {
+    // This seems to be the fastest way of doing this without using a bunch of unsafe:
+    // https://gist.github.com/thomcc/c4860d68cf31f9b0283c692f83a239f3
+    static HEX_CHARS: &'static [u8] = b"0123456789abcdef";
+    let mut result = vec![0u8; bytes.len() * 2];
+    let mut index = 0;
+    for &byte in bytes {
+        result[index + 0] = HEX_CHARS[(byte >> 4) as usize];
+        result[index + 1] = HEX_CHARS[(byte & 15) as usize];
+        index += 2;
+    }
+    // We know statically that this unwrap is safe, since we can only write ascii
+    String::from_utf8(result).unwrap()
+}
+
+pub fn random_guid() -> Result<String, openssl::error::ErrorStack> {
+    let mut bytes = vec![0u8; 9];
+    openssl::rand::rand_bytes(&mut bytes)?;
+    Ok(base64::encode_config(&bytes, base64::URL_SAFE_NO_PAD))
+}
+
+/// Typesafe way to manage server timestamps without accidentally mixing them up with
+/// local ones.
+///
+/// TODO: We should probably store this as milliseconds (or something) for stability and to get
+/// Eq/Ord. The server guarantees that these are formatted to the hundreds place (not sure if this
+/// is documented but the code does it intentionally...). This would also let us throw out negative
+/// and NaN timestamps, which the server certainly won't send, but the guarantee would make me feel
+/// better.
+#[derive(Debug, Copy, Clone, PartialEq, PartialOrd, Deserialize, Serialize, Default)]
+pub struct ServerTimestamp(pub f64);
```

Comment:
this is very cool - I wonder if we should have something similar for durations? (I think mentat has something similar)

Reply:
Well, `std::time::Duration` exists. But yeah I was thinking we might want something for the kinds of durations the server sends, which are in floating point seconds.